contract/src/main.rs

```diff
@@ -807,54 +813,61 @@ pub extern "C" fn approve() {
 
     if let NFTIdentifierMode::Ordinal = identifier_mode {
         // Revert if token_id is out of bounds
-        if let TokenIdentifier::Index(index) = &token_identifier {
+        if let TokenIdentifier::Index(index) = &token_id {
             if *index >= number_of_minted_tokens {
                 runtime::revert(NFTCoreError::InvalidTokenIdentifier);
             }
         }
     }
 
-    let token_owner_key = match utils::get_dictionary_value_from_key::<Key>(
+    let owner = match utils::get_dictionary_value_from_key::<Key>(
         TOKEN_OWNERS,
         &token_identifier_dictionary_key,
     ) {
-        Some(token_owner) => token_owner,
+        Some(owner) => owner,
         None => runtime::revert(NFTCoreError::InvalidAccountHash),
     };
 
-    // Revert if caller is not the token_owner. Only the token owner can approve an operator
-    if token_owner_key != caller {
+    // Revert if caller is not token owner nor operator.
+    // Only the token owner or an operator can approve an account
+    let is_owner = caller == owner;
+    let is_operator = !is_owner
+        && match utils::get_dictionary_value_from_key::<Option<Key>>(
+            OPERATORS,
+            &utils::get_stringified_token_owner_item_key(owner),
+        ) {
+            Some(Some(maybe_operator)) => caller == maybe_operator,
+            Some(None) | None => false,
+        };
+
+    if !is_owner && !is_operator {
         runtime::revert(NFTCoreError::InvalidAccountHash);
     }
 
     // We assume a burnt token cannot be approved
-    if utils::is_token_burned(&token_identifier) {
+    if utils::is_token_burned(&token_id) {
         runtime::revert(NFTCoreError::PreviouslyBurntToken)
     }
 
-    let operator = utils::get_named_arg_with_user_errors::<Key>(
-        ARG_OPERATOR,
-        NFTCoreError::MissingApprovedAccountHash,
-        NFTCoreError::InvalidApprovedAccountHash,
-    )
-    .unwrap_or_revert();
+    let spender = match utils::get_optional_named_arg_with_user_errors::<Key>(
+        ARG_OPERATOR, // Deprecated in favor of ARG_SPENDER
+        NFTCoreError::MissingContractWhiteList,
```

Comment:
Thanks please check that [commit](https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/181/commits/c632e91fd84f7e4927cd01145d2027b894d25c2d), MissingApprovedAccountHash gets deprecated because ARG_OPERATOR becomes optional. Adding two new errors for explicit spender errors and fixing that copy-paste typo.